### PR TITLE
AD9545: Add driver and basic project example

### PR DESCRIPTION
**Pull request by jonathanns**
## Pull Request Description

Add support for AD9545 Quad Input, 10-Output, Dual DPLL/IEEE 1588, 1 pps Synchronizer and Jitter Cleaner. Also includes a Basic project example tested in AD-SYNCHRONA14-EBZ.

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [x] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [ ] I have build all projects affected by the changes in this PR
- [ ] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [ ] I have updated the documentation (wiki pages, ReadMe etc), if applies

### CODE REVIEW

**Comment by jonathanns** [comment]
This is a WIP in the final stages.
Some dependencies are available on #2268.

File: drivers/frequency/ad9545/ad9545.h [hunk 1]
```diff
@@ -0,0 +1,826 @@
...
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+#ifndef ad9545_H_
```
Comment (danmois) [inline drivers/frequency/ad9545/ad9545.h:40]: Caps?

File: drivers/frequency/ad9545/ad9545.h [hunk 2]
```diff
@@ -0,0 +1,826 @@
...
+	uint8_t		tdc_source;
+	bool		fb_tagging;
+};
+
+struct ad9545_pll_clk {
+	struct ad9545_dev		*dev;
+	bool				pll_used;
+	unsigned int			address;
+	struct no_os_clk_desc		*hw;
+	uint8_t			num_parents;
```
Comment (danmois) [inline drivers/frequency/ad9545/ad9545.h:648]: Alignment

File: drivers/frequency/ad9545/ad9545.c [hunk 3]
```diff
@@ -0,0 +1,2434 @@
...
+
+static int ad9545_get_r_div(struct ad9545_dev *dev, int addr, uint32_t *r_div)
+{
+	uint32_t regval;
+	uint32_t val, div;
+	int ret;
+
+	ret = ad9545_read_reg_multiple(dev, AD9545_REF_X_RDIV(addr), (uint8_t *)&regval,
+				       4);
+	if (ret < 0)
```
Comment (danmois) [inline drivers/frequency/ad9545/ad9545.c:160]: if (ret) ?

File: drivers/frequency/ad9545/ad9545.c [hunk 4]
```diff
@@ -0,0 +1,2434 @@
...
+	.clk_disable = ad9545_out_clk_disable,
+	.clk_recalc_rate = ad9545_out_clk_recalc_rate,
+	.clk_round_rate = ad9545_out_clk_round_rate,
+	.clk_set_rate = ad9545_out_clk_set_rate,
+};
+
+
+static int32_t ad9545_check_id(struct ad9545_dev *dev)
+{
+
```
Comment (danmois) [inline drivers/frequency/ad9545/ad9545.c:1240]: Extra line

File: drivers/frequency/ad9545/ad9545.c [hunk 5]
```diff
@@ -0,0 +1,2434 @@
...
+		pr_err("Unrecognized CHIP_ID 0x%X\n", chip_id);
+		return -ENODEV;
+	}
+	return ret;
+}
+
+static int ad9545_set_r_div(struct ad9545_dev *dev, uint32_t div, int addr)
+{
+	uint32_t regval;
+	uint32_t val;
```
Comment (danmois) [inline drivers/frequency/ad9545/ad9545.c:1265]: Is val required in this function?

File: drivers/frequency/ad9545/ad9545.c [hunk 6]
```diff
@@ -0,0 +1,2434 @@
...
+ */
+int32_t ad9545_init(struct ad9545_dev **device,
+		    struct ad9545_init_param *init_param)
+{
+	struct ad9545_dev *dev;
+	int32_t ret;
+
+	dev = (struct ad9545_dev *)no_os_malloc(sizeof(*dev));
+	if (!dev)
+		goto error;
```
Comment (danmois) [inline drivers/frequency/ad9545/ad9545.c:1553]: What will we return from here?

File: drivers/frequency/ad9545/ad9545.c [hunk 7]
```diff
@@ -0,0 +1,2434 @@
...
+	}
+
+	/* Check device */
+	ret = ad9545_check_id(dev);
+	if (ret < 0)
+		goto error;
+
+	/* Device settings */
+	ret = ad9545_parse_init(dev, init_param);
+	if (!ret) {
```
Comment (danmois) [inline drivers/frequency/ad9545/ad9545.c:1596]: Looks somehow different than what we have. I'm not saying that it is not correct, only that I am used with the flow below:
if (ret)
   goto error;

*device = dev; ...

File: drivers/frequency/ad9545/ad9545.c [hunk 8]
```diff
@@ -0,0 +1,2434 @@
...
+	fosc = NO_OS_DIV_ROUND_UP(dev->sys_clk.ref_freq_hz, 1000000);
+
+	if (dev->sys_clk.sys_clk_freq_doubler)
+		fosc *= 2;
+
+	div_ratio = 0;
+	for (i = 4; i < 256; i++) {
+		fs = i * fosc;
+
+		if (fs > 2250 && fs < 2415) {
```
Comment (danmois) [inline drivers/frequency/ad9545/ad9545.c:1637]: Could we have macros here?

File: drivers/frequency/ad9545/ad9545.c [hunk 9]
```diff
@@ -0,0 +1,2434 @@
...
+			ret = ad9545_write_reg(dev, AD9545_PWR_CALIB_CHX(i), 0);
+			if (ret < 0)
+				return ret;
+
+			ret = ad9545_io_update(dev);
+			if (ret < 0)
+				return ret;
+
+			cal_count = 2;
+			break;
```
Comment (danmois) [inline drivers/frequency/ad9545/ad9545.c:2048]: Is break required here?

File: projects/ad9545/src/examples/basic/basic_example.c [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+		      200000);
+	if (ret)
+		goto error;
+
+	/* Recalibrate PLLs */
+	ret = ad9545_calib_aplls(ad9545_dev);
+	if (ret < 0)
+		return ret;
+
+	while(1) {
```
Comment (danmois) [inline projects/ad9545/src/examples/basic/basic_example.c:267]: the program could finish with a "Done" message if everything was ok. Do we need the loop here after the device is setup?

File: drivers/frequency/ad9545/ad9545.h (hunk 1 above)
Comment (jonathanns) [reply drivers/frequency/ad9545/ad9545.h:40]: noted

File: drivers/frequency/ad9545/ad9545.h (hunk 2 above)
Comment (jonathanns) [reply drivers/frequency/ad9545/ad9545.h:648]: noted

File: drivers/frequency/ad9545/ad9545.h [hunk 11]
```diff
@@ -0,0 +1,826 @@
...
+	bool				sys_clk_freq_doubler;
+	bool				sys_clk_crystal;
+	uint32_t			ref_freq_hz;
+	uint32_t			sys_freq_hz;
+};
+
+struct ad9545_dev;
+
+typedef int32_t (*ad9545_reg_read_func)(struct ad9545_dev *dev,
+					uint16_t reg_addr,
```
[the remark this answers is not in the document]
Comment (jonathanns) [reply drivers/frequency/ad9545/ad9545.h:721]: noted

File: drivers/frequency/ad9545/ad9545.h [hunk 12]
```diff
@@ -0,0 +1,826 @@
...
+				ad9545_aux_nco_clk_names)];
+	struct ad9545_aux_tdc_clk	aux_tdc_clks[NO_OS_ARRAY_SIZE(
+				ad9545_aux_tdc_clk_names)];
+
+};
+
+/******************************************************************************/
+/************************ Functions Declarations ******************************/
+/******************************************************************************/
+int32_t ad9545_spi_reg_read(struct ad9545_dev *dev,
```
[the remark this answers is not in the document]
Comment (jonathanns) [reply drivers/frequency/ad9545/ad9545.h:783]: ok!

File: drivers/frequency/ad9545/ad9545.c (hunk 3 above)
Comment (jonathanns) [reply drivers/frequency/ad9545/ad9545.c:160]: Sure, i can replace.

File: drivers/frequency/ad9545/ad9545.c [hunk 13]
```diff
@@ -0,0 +1,2434 @@
...
+	.clk_set_rate = ad9545_pll_set_rate,
+};
+
+static int ad9545_get_nco_center_freq(struct ad9545_dev *dev, int addr,
+				      uint64_t *freq)
+{
+	uint64_t regval64;
+	int ret;
+
+	regval64 = 0;
```
[the remark this answers is not in the document]
Comment (jonathanns) [reply drivers/frequency/ad9545/ad9545.c:769]: ok

File: drivers/frequency/ad9545/ad9545.c (hunk 5 above)
Comment (jonathanns) [reply drivers/frequency/ad9545/ad9545.c:1265]: Actually not. removing...

File: drivers/frequency/ad9545/ad9545.c (hunk 6 above)
Comment (jonathanns) [reply drivers/frequency/ad9545/ad9545.c:1553]: Actually it should be:
```
if (!dev) 
           return -ENOMEM;
```
if you agree.

File: drivers/frequency/ad9545/ad9545.c (hunk 7 above)
Comment (jonathanns) [reply drivers/frequency/ad9545/ad9545.c:1596]: Sure, i agree with you.

File: drivers/frequency/ad9545/ad9545.c (hunk 8 above)
Comment (jonathanns) [reply drivers/frequency/ad9545/ad9545.c:1637]: noted

File: projects/ad9545/src/examples/basic/basic_example.c (hunk 10 above)
Comment (jonathanns) [reply projects/ad9545/src/examples/basic/basic_example.c:267]: Yes, indeed the loop is not necessary, maybe i left there after the tests, i will remove.

File: drivers/frequency/ad9545/ad9545.c (hunk 9 above)
Comment (jonathanns) [reply drivers/frequency/ad9545/ad9545.c:2048]: In this case it seems redundant with the previous line.

**Comment by jonathanns** [comment]
Update v1: 
* Fixed nits.
* Added brief descriptions to function headers.
* Replaced constant values with macros.
* Removed unnecessary variables.

File: projects/ad9545/src/examples/basic/basic_example.c [hunk 14]
```diff
@@ -0,0 +1,274 @@
...
+	if (ret)
+		goto error;
+
+	/* Recalibrate PLLs */
+	ret = ad9545_calib_aplls(ad9545_dev);
+	if (ret < 0)
+		return ret;
+
+	printf("Done.\n\r");
+	return 0;
```
Comment (danmois) [inline projects/ad9545/src/examples/basic/basic_example.c:268]: We should also remove the device descriptor before exiting the program.

**Comment by jonathanns** [comment]
Update v2: 
* Fixed the `ad9545_remove` function.
* Added device descriptor removal before exiting program.

**Review by danmois — APPROVED** [review]
lgtm

**Review by amiclaus — APPROVED** [review]

**Pull request merged**